relay-server/src/actors/processor.rs

```diff
@@ -850,27 +850,24 @@ impl EnvelopeProcessor {
 
     /// Remove profiles if the feature flag is not enabled
     fn process_profiles(&self, state: &mut ProcessEnvelopeState) {
+        let profiling_enabled = state.project_state.has_feature(Feature::Profiling);
+        let envelope = &mut state.envelope;
+
+        envelope.retain_items(|item| match item.ty() {
+            ItemType::Profile => profiling_enabled,
+            _ => true,
+        });
+
         if !self.config.processing_enabled() {
             return;
         }
 
-        let profiling_enabled = state.project_state.has_feature(Feature::Profiling);
-        let envelope = &mut state.envelope;
         let context = &state.envelope_context;
+        let new_profiles: &mut Vec<Vec<u8>> = &mut Vec::new();
 
-        if let Some(item) = envelope.take_item_by(|item| item.ty() == &ItemType::Profile) {
-            if !profiling_enabled {
-                return;
-            }
-
+        while let Some(item) = envelope.take_item_by(|item| item.ty() == &ItemType::Profile) {
             match relay_profiling::expand_profile(&item.payload()[..]) {
-                Ok(payloads) => {
-                    for payload in payloads.iter() {
-                        let mut item = Item::new(ItemType::Profile);
-                        item.set_payload(ContentType::Json, &payload[..]);
-                        envelope.add_item(item);
-                    }
-                }
+                Ok(mut payloads) => new_profiles.append(&mut payloads),
```

Comment:
What's the advantage of introducing the intermediate vector `new_profiles` here, instead of calling `envelope.add_item` directly?

Reply:
That would create an endless loop in the outer loop, since that runs until there's no more profiling item in the envelope.